File: go/objectmodel/node.go
package objectmodel

import (
	"strconv"
	"strings"
)

// valueKind is the arity of a node's payload (SPEC §2.1, `Value`).
type valueKind int

const (
	// kindAbsent is a staging state only: the node has no value yet. SPEC
	// §8.5 either fills it or the node does not exist.
	kindAbsent valueKind = iota
	kindScalar
	kindList
	kindMap
	kindOpaque
	// kindRaw is payload kept verbatim: an opaque value, or a list inside a
	// primitive declaration, which declares no element position and so is not
	// materialized into nodes (INV-035). Scalar leaves are kindScalar; the
	// mapping levels above them are kindMap.
	kindRaw
)

// Node is a CIC node (SPEC §2.1): exactly one `values` member (INV-001) and,
// in canonical form, exactly one `origin` member (INV-002), plus one member
// per primitive its schema declares (INV-022).
//
// Every field is unexported and there is no exported constructor, so a Node
// can only come out of Materialize. That is half of INV-032; CanonicalObject
// in materialize.go is the other half.
type Node struct {
	path   string
	isRoot bool
	eff    *effNode

	kind    valueKind
	scalar  any
	list    []*Node
	entries map[string]*Node
	order   []string
	opaque  any
	raw     any

	origin Origin

	// fromAbsence marks an object node that no author and no template
	// asserted. If nothing below it materializes, it does not exist.
	fromAbsence bool

	primitives map[string]*Node
	primOrder  []string

	// staging, populated by §8.4 and consumed by §8.6
	authoredPrims map[string]any
	extraMembers  map[string]any
}

// Path returns the node's address in the object ($.interface.mtu).
func (n *Node) Path() string { return n.path }

// Origin returns the node's authoring authority (SPEC §5), copied.
//
// The copy is the whole point. This method used to return the node's own
// slice, so a caller could write through it and change the provenance of a
// validated object in place:
//
//	obj.Root().Origin()[0].Kind = "evil"
//
// After that the tree said `evil` while CanonicalYAML() still said `schema`,
// byte for byte unchanged. No forgery and no unsafe was needed — one
// assignment split the two views of one object apart, and which half a reader
// consults then decides what it believes about who authored the value.
//
// Children() and Primitives() already copied; this one was missed. OriginTerm
// holds only strings, so copying the slice copies everything reachable.
func (n *Node) Origin() Origin {
	out := make(Origin, len(n.origin))
	copy(out, n.origin)
	return out
}

// Child returns a child node of a structured payload.
func (n *Node) Child(name string) (*Node, bool) {
	if n.kind != kindMap {
		return nil, false
	}
	c, ok := n.entries[name]
	return c, ok
}

// Primitive returns a materialized primitive member of the node.
func (n *Node) Primitive(name string) (*Node, bool) {
	p, ok := n.primitives[name]
	return p, ok
}

// Scalar returns a scalar payload.
func (n *Node) Scalar() (any, bool) {
	if n.kind != kindScalar {
		return nil, false
	}
	return n.scalar, true
}

// Len reports the number of entries in a list payload, or 0 for any other kind.
func (n *Node) Len() int {
	if n.kind != kindList {
		return 0
	}
	return len(n.list)
}

// At returns the i-th entry of a list payload.
//
// A list entry is a node when the schema declared an element position via
// `item:` (INV-035). Where it did not — a CertPattern list inside an access
// rule, for instance — the list is the payload of its own node and has no
// entries to walk here; Len reports 0 for it.
func (n *Node) At(i int) (*Node, bool) {
	if n.kind != kindList || i < 0 || i >= len(n.list) {
		return nil, false
	}
	return n.list[i], true
}

// Children returns the child names of a structured payload, in canonical order.
func (n *Node) Children() []string {
	if n.kind != kindMap {
		return nil
	}
	out := make([]string, len(n.order))
	copy(out, n.order)
	return out
}

// Primitives returns the names of the node's materialized primitives, in the
// order SPEC §6.1 states the atom set.
func (n *Node) Primitives() []string {
	out := make([]string, len(n.primOrder))
	copy(out, n.primOrder)
	return out
}

// Get resolves an address against this node (SPEC §2.5) and is the operation
// §2.2 promises: every declared position is a first-class, addressable object. Without it the claim is true of the data and unavailable to a caller,
// who has to hand-walk Child and Primitive to reach a position the model says
// is addressable.
//
// The syntax is the one Error.Path already uses, so an address read out of a
// rejection can be fed straight back in:
//
//	$.values.mtu.access               absolute — `mtu`'s access primitive
//	$.values.mtu.access.values.read   the access payload's `read` child
//	values.mtu                        relative to this node
//	$.values.addresses.values[0]      a list entry
//
// `values` steps into the payload and is never optional; anything else is a
// member of the node itself. That is the whole grammar, and it is what makes
// an address unique (INV-040): without the mandatory step, `$.values.mtu` and
// `$.values.values.values.mtu` would name the same node, and a primitive would
// be indistinguishable from a payload child that shares its name.
//
// Get does not create anything and reports false for an address that does not
// resolve. It is a reader, not a cursor.
func (n *Node) Get(path string) (*Node, bool) {
	cur := n
	inPayload := false // did the previous segment step into the payload?

	for _, seg := range splitPath(path) {
		switch {
		case inPayload:
			// Only a payload child can follow a `values` step.
			c, ok := cur.Child(seg)
			if !ok {
				return nil, false
			}
			cur, inPayload = c, false

		case seg == "values":
			// The payload step itself resolves nothing; the segment after it
			// does. Keeping it as state rather than a node is what stops
			// `$.values.values.mtu` from meaning `$.values.mtu`.
			inPayload = true

		default:
			if i, ok := parseIndex(seg); ok {
				// `values[i]` is the payload step and the entry in one token.
				e, ok := cur.At(i)
				if !ok {
					return nil, false
				}
				cur = e
				continue
			}
			// Anything else on a node is a member of the node: a primitive.
			p, ok := cur.Primitive(seg)
			if !ok {
				return nil, false
			}
			cur = p
		}
	}
	if inPayload {
		// A trailing `values` names the payload, which is not a node.
		return nil, false
	}
	return cur, true
}

// parseIndex recognises the `values[i]` token of SPEC §2.5.
func parseIndex(seg string) (int, bool) {
	if !strings.HasPrefix(seg, "values[") || !strings.HasSuffix(seg, "]") {
		return 0, false
	}
	i, err := strconv.Atoi(seg[len("values[") : len(seg)-1])
	if err != nil {
		return 0, false
	}
	return i, true
}

// splitPath turns an address into segments, tolerating the leading `$` that
// Path() and Error.Path carry. `values[0]` stays one segment.
func splitPath(path string) []string {
	path = strings.TrimPrefix(path, "$")
	path = strings.TrimPrefix(path, ".")
	if path == "" {
		return nil
	}
	return strings.Split(path, ".")
}

func (n *Node) setPrimitive(name string, p *Node) {
	if n.primitives == nil {
		n.primitives = map[string]*Node{}
	}
	if _, ok := n.primitives[name]; !ok {
		n.primOrder = append(n.primOrder, name)
	}
	n.primitives[name] = p
}

File: rust/src/node.rs
//! The canonical node and the reader API.
//!
//! A `Node` cannot be built outside this crate: every field is private and
//! there is no public constructor, so the only way to obtain one is
//! `materialize`. That is INV-032, enforced by the type system rather than by
//! asking callers to behave.
//!
//! Note what Rust gives here for free that Go did not. `Origin()` in the Go
//! implementation handed out the node's own slice, and a caller could rewrite
//! the provenance of a validated object through it while the serialization
//! stayed byte-identical (audit finding F-03). Here `origin()` returns `&Origin`
//! — a shared borrow of an immutable value — so the same mistake does not
//! compile. This is not the Rust implementation being more careful; it is the
//! language making one class of the bug unrepresentable.

use crate::origin::Origin;
use crate::value::Value;

/// A node's payload — the `values` member (INV-001).
#[derive(Debug, Clone, PartialEq)]
pub enum Payload {
    /// A single value.
    Scalar(Value),
    /// A list whose element positions the schema declared, so each entry is a
    /// node in its own right.
    Seq(Vec<Node>),
    /// A structured payload: named children, in canonical order.
    Map(Vec<(String, Node)>),
    /// Kept verbatim. An opaque payload (SPEC §4), or a list inside a primitive
    /// declaration, which declares no element position and so is not
    /// materialized into nodes (INV-035).
    Raw(Value),
}

/// A CIC node: exactly one `values` member (INV-001), exactly one `origin`
/// member (INV-002), and one member per primitive its schema declares
/// (INV-022).
#[derive(Debug, Clone, PartialEq)]
pub struct Node {
    pub(crate) path: String,
    pub(crate) is_root: bool,
    pub(crate) payload: Payload,
    pub(crate) origin: Origin,
    /// Primitive members, already in the §6.1 canonical order.
    pub(crate) primitives: Vec<(String, Node)>,
}

impl Node {
    /// This node's address in the object (`$.values.mtu`).
    #[must_use]
    pub fn path(&self) -> &str {
        &self.path
    }

    /// The node's authoring authority (SPEC §5).
    #[must_use]
    pub fn origin(&self) -> &Origin {
        &self.origin
    }

    /// The payload, for callers that need to match on its arity.
    #[must_use]
    pub fn payload(&self) -> &Payload {
        &self.payload
    }

    /// A scalar payload, if that is what this node has.
    #[must_use]
    pub fn scalar(&self) -> Option<&Value> {
        match &self.payload {
            Payload::Scalar(v) => Some(v),
            _ => None,
        }
    }

    /// A named child of a structured payload.
    #[must_use]
    pub fn child(&self, name: &str) -> Option<&Node> {
        match &self.payload {
            Payload::Map(entries) => entries.iter().find(|(n, _)| n == name).map(|(_, c)| c),
            _ => None,
        }
    }

    /// Child names, canonical order.
    #[must_use]
    pub fn children(&self) -> Vec<&str> {
        match &self.payload {
            Payload::Map(entries) => entries.iter().map(|(n, _)| n.as_str()).collect(),
            _ => Vec::new(),
        }
    }

    /// Entry count of a list payload; zero for anything else.
    #[must_use]
    pub fn len(&self) -> usize {
        match &self.payload {
            Payload::Seq(items) => items.len(),
            _ => 0,
        }
    }

    #[must_use]
    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }

    /// A list entry by position.
    #[must_use]
    pub fn at(&self, i: usize) -> Option<&Node> {
        match &self.payload {
            Payload::Seq(items) => items.get(i),
            _ => None,
        }
    }

    /// A materialized primitive member.
    #[must_use]
    pub fn primitive(&self, name: &str) -> Option<&Node> {
        self.primitives
            .iter()
            .find(|(n, _)| n == name)
            .map(|(_, p)| p)
    }

    /// Primitive names, canonical order.
    #[must_use]
    pub fn primitive_names(&self) -> Vec<&str> {
        self.primitives.iter().map(|(n, _)| n.as_str()).collect()
    }

    /// Resolve an address (SPEC §2.5).
    ///
    /// `values` steps into the payload and is never optional; anything else is
    /// a member of the node itself. That mandatory step is what makes an
    /// address unique (INV-040): without it `$.values.mtu` and
    /// `$.values.values.values.mtu` would name the same node, and a primitive
    /// would be indistinguishable from a payload child sharing its name.
    ///
    /// A leading `$` means the root and is not decoration — an absolute address
    /// is answered by the root and by nothing else. Without it the address is
    /// relative to the receiver. Malformed spellings are rejected rather than
    /// repaired, so a caller can tell whether the string it holds is canonical.
    #[must_use]
    pub fn get(&self, path: &str) -> Option<&Node> {
        let (segments, absolute) = parse_path(path)?;
        if absolute && !self.is_root {
            return None;
        }

        let mut cur = self;
        let mut in_payload = false;
        for seg in segments {
            if in_payload {
                cur = cur.child(seg)?;
                in_payload = false;
            } else if seg == "values" {
                in_payload = true;
            } else if let Some(i) = parse_index(seg) {
                cur = cur.at(i)?;
            } else {
                cur = cur.primitive(seg)?;
            }
        }
        // A trailing `values` names the payload, which is not a node.
        if in_payload {
            return None;
        }
        Some(cur)
    }
}

/// `values[i]` — the payload step and the entry in one token.
fn parse_index(seg: &str) -> Option<usize> {
    seg.strip_prefix("values[")?
        .strip_suffix(']')?
        .parse::<usize>()
        .ok()
}

/// Split an address, reporting whether it is absolute. Returns `None` for a
/// malformed address.
fn parse_path(path: &str) -> Option<(Vec<&str>, bool)> {
    match path {
        "" => return Some((Vec::new(), false)), // the receiver
        "$" => return Some((Vec::new(), true)), // the root
        _ => {}
    }
    let (rest, absolute) = match path.strip_prefix('$') {
        // `$values.mtu` — `$` not followed by a separator.
        Some(r) => (r.strip_prefix('.')?, true),
        None => (path, false),
    };
    if rest.is_empty() || rest.starts_with('.') || rest.ends_with('.') {
        return None;
    }
    let segments: Vec<&str> = rest.split('.').collect();
    if segments.iter().any(|s| s.is_empty()) {
        return None;
    }
    Some((segments, absolute))
}
